### PR TITLE
add: new policy to mutate dnsPolicy and dnsConfig

### PR DESCRIPTION
**Pull request by ugur99**
Signed-off-by: Ugur Can Ozturk <ugurozturk918@gmail.com>

## Related Issue(s)

None

## Description

The Default DNS policy in Kubernetes gives much flexibility in service access/discovery; however, it costs some latency and performance problems on high scale, and it needs to be optimized. This policy helps us to optimize the performance of DNS queries by eliminating many dummy DNS requests by setting more strict and optimal options for `dnsConfig` and `dnsPolicy`.

## Checklist

- [x] I have read the [policy contribution guidelines](https://github.com/kyverno/policies/blob/main/README.md#contribution).
- [x] I have added test manifests and resources covering both positive and negative tests that prove this policy works as intended.

### CODE REVIEW

**Review by chipzoller — CHANGES_REQUESTED** [review]
Request you to please add tests (positive and negative) for your sample here.

File: other/dns-policy-and-dns-config/dns-policy-and-dns-config.yaml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    context:
+    - name: dictionary
+      configMap:
+        # kubelet-config cm would also works by using clusterDomain 
+        # instead of clusterName; but kubeadm-config sounds more reliable
+        # when considering kubelet-config is changed every cluster upgrade, etc.
+        name: kubeadm-config 
+        namespace: kube-system
+    match:
+      resources:
```
Comment (chipzoller) [inline other/dns-policy-and-dns-config/dns-policy-and-dns-config.yaml:29]: Please add an `any` or `all` block under `match` here to use the latest policy definition language.

File: other/dns-policy-and-dns-config/dns-policy-and-dns-config.yaml [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        kinds:
+        - Pod
+    preconditions:
+      any:
+      - key: "{{ request.object.spec.dnsPolicy }}"
+        operator: In
```
Comment (chipzoller) [inline other/dns-policy-and-dns-config/dns-policy-and-dns-config.yaml:35]: `AnyIn` should be the operator.

File: other/dns-policy-and-dns-config/dns-policy-and-dns-config.yaml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+            options:
+            - name: timeout
+              value: "1"
+            - name: ndots
+              value: "2"
+            - name: attempts
+              value: "1"
+            searches:
+            - svc.{{dictionary.data.ClusterConfiguration | parse_yaml(@).clusterName}}
+            - '{{ request.namespace }}.svc.{{ dictionary.data.ClusterConfiguration | parse_yaml(@).clusterName }}'
```
Comment (chipzoller) [inline other/dns-policy-and-dns-config/dns-policy-and-dns-config.yaml:56]: Best to wrap this line in double quotes rather than single.

File: other/dns-policy-and-dns-config/dns-policy-and-dns-config.yaml [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+spec:
+  rules:
+  - name: dns-policy
+    context:
+    - name: dictionary
+      configMap:
+        # kubelet-config cm would also works by using clusterDomain 
+        # instead of clusterName; but kubeadm-config sounds more reliable
+        # when considering kubelet-config is changed every cluster upgrade, etc.
+        name: kubeadm-config
```
Comment (chipzoller) [inline other/dns-policy-and-dns-config/dns-policy-and-dns-config.yaml:26]: Could you add an example ConfigMap to this policy dir as well?

File: other/dns-policy-and-dns-config/dns-policy-and-dns-config.yaml (hunk 1 above)
Comment (ugur99) [reply other/dns-policy-and-dns-config/dns-policy-and-dns-config.yaml]: It's done, I added `any` block.

File: other/dns-policy-and-dns-config/dns-policy-and-dns-config.yaml (hunk 2 above)
Comment (ugur99) [reply other/dns-policy-and-dns-config/dns-policy-and-dns-config.yaml]: Changed Operator to `AnyIn` :)

File: other/dns-policy-and-dns-config/dns-policy-and-dns-config.yaml (hunk 3 above)
Comment (ugur99) [reply other/dns-policy-and-dns-config/dns-policy-and-dns-config.yaml]: double quotes replaced with single ones

File: other/dns-policy-and-dns-config/dns-policy-and-dns-config.yaml (hunk 4 above)
Comment (ugur99) [reply other/dns-policy-and-dns-config/dns-policy-and-dns-config.yaml:26]: I added an example of kubeadm-config configmap

**Comment by ugur99** [comment]
@chipzoller Thanks for your review; finally I figured out to pass variables to the test. I added a test scenario in the folder. 


```
Executing dns-policy...
applying 1 policy to 1 resource... 

│───│────────────│────────────│───────────────────────│────────│
│ # │ POLICY     │ RULE       │ RESOURCE              │ RESULT │
│───│────────────│────────────│───────────────────────│────────│
│ 1 │ dns-policy │ dns-policy │ default/Pod/myapp-pod │ Pass   │
│───│────────────│────────────│───────────────────────│────────│

Test Summary: 1 tests passed and 0 tests failed
```

**Comment by ugur99** [comment]
Hi @chipzoller, are there any remediations you suggest that I need to do?

**Comment by chipzoller** [comment]
Hi, don't think so, but we'll have to take a closer look after KubeCon most likely. Thanks for this contribution!

File: other/dns-policy-and-dns-config/dns-policy-and-dns-config.yaml [hunk 5]
```diff
@@ -0,0 +1,58 @@
+apiVersion: kyverno.io/v1
+kind: ClusterPolicy
+metadata:
+  name: dns-policy
```
Comment (chipzoller) [inline other/dns-policy-and-dns-config/dns-policy-and-dns-config.yaml:4]: Can you update the name to be closer to the title annotation? Like maybe `change-dns-config-policy` or similar?

File: other/dns-policy-and-dns-config/dns-policy-and-dns-config.yaml [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  annotations:
+    policies.kyverno.io/title: Change DNS Config and Policy
+    policies.kyverno.io/category: Other
+    policies.kyverno.io/severity: medium
+    kyverno.io/kyverno-version: 1.7.3
```
Comment (chipzoller) [inline other/dns-policy-and-dns-config/dns-policy-and-dns-config.yaml:9]: This annotation holds the version of Kyverno you actually tested and verified operation. Is that actually 1.7.3? Can you check and certify on Kyverno 1.8.1?

File: other/dns-policy-and-dns-config/dns-policy-and-dns-config.yaml [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    kyverno.io/kubernetes-version: "1.23"
```
Comment (chipzoller) [inline other/dns-policy-and-dns-config/dns-policy-and-dns-config.yaml:10]: Same with Kubernetes version.

File: other/dns-policy-and-dns-config/dns-policy-and-dns-config.yaml [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    policies.kyverno.io/subject: Pod
+    policies.kyverno.io/description: >-
```
Comment (chipzoller) [inline other/dns-policy-and-dns-config/dns-policy-and-dns-config.yaml:12]: You may also want to state this policy requires a kubeadm-provisioned cluster to operate properly. Remember that not all clusters are built using kubeadm or even have the same resources available.

**Review by chipzoller — CHANGES_REQUESTED** [review]

File: other/dns-policy-and-dns-config/dns-policy-and-dns-config.yaml (hunk 6 above)
Comment (ugur99) [reply other/dns-policy-and-dns-config/dns-policy-and-dns-config.yaml]: Actually, it works with both 1.7.3 and 1.8.1; I checked it.

File: other/dns-policy-and-dns-config/dns-policy-and-dns-config.yaml (hunk 7 above)
Comment (ugur99) [reply other/dns-policy-and-dns-config/dns-policy-and-dns-config.yaml:10]: I tested it on the cluster with the 1.23 version.

File: other/dns-policy-and-dns-config/dns-policy-and-dns-config.yaml (hunk 8 above)
Comment (ugur99) [reply other/dns-policy-and-dns-config/dns-policy-and-dns-config.yaml:12]: Yes, you are right, I fixed the description.

**Review by chipzoller — APPROVED** [review]
